azurerm/internal/services/blueprints/blueprint.go

```diff
@@ -0,0 +1,211 @@
+						Type:         schema.TypeString,
+						ValidateFunc: validate.UserAssignedIdentityId,
+					},
+				},
+
+				"principal_id": {
+					Type:     schema.TypeString,
+					Computed: true,
+				},
+
+				"tenant_id": {
+					Type:     schema.TypeString,
+					Computed: true,
+				},
+			},
+		},
+	}
+}
+
+func blueprintAssignmentCreateStateRefreshFunc(ctx context.Context, client *blueprint.AssignmentsClient, scope, name string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		resp, err := client.Get(ctx, scope, name)
+		if err != nil {
+			return nil, "", fmt.Errorf("unable to retrieve Blueprint Assignment %q (Scope %q): %+v", name, scope, err)
+		}
+		if resp.ProvisioningState == blueprint.Failed {
+			return resp, string(resp.ProvisioningState), err
+		}
+
+		return resp, string(resp.ProvisioningState), nil
+	}
+}
+
+func blueprintAssignmentDeleteStateRefreshFunc(ctx context.Context, client *blueprint.AssignmentsClient, scope, name string) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		resp, err := client.Get(ctx, scope, name)
+		if err != nil {
+			if utils.ResponseWasNotFound(resp.Response) {
+				return resp, "NotFound", nil
+			} else {
+				return nil, "", fmt.Errorf("unable to retrieve Blueprint Assignment %q (Scope %q): %+v", name, scope, err)
+			}
+		}
+
+		return resp, string(resp.ProvisioningState), nil
+	}
+}
+
+func normalizeAssignmentParameterValuesJSON(jsonString interface{}) string {
+	if jsonString == nil || jsonString == "" {
+		return ""
+	}
+
+	var values map[string]*blueprint.ParameterValue
+	if err := json.Unmarshal([]byte(jsonString.(string)), &values); err != nil {
+		return fmt.Sprintf("unable to parse JSON: %+v", err)
+	}
+
+	b, _ := json.Marshal(values)
+	return string(b)
+}
+
+func normalizeAssignmentResourceGroupValuesJSON(jsonString interface{}) string {
+	if jsonString == nil || jsonString == "" {
+		return ""
+	}
+
+	var values map[string]*blueprint.ResourceGroupValue
+	if err := json.Unmarshal([]byte(jsonString.(string)), &values); err != nil {
+		return fmt.Sprintf("unable to parse JSON: %+v", err)
+	}
+
+	b, _ := json.Marshal(values)
+	return string(b)
+}
+
+func expandArmBlueprintAssignmentParameters(input string) map[string]*blueprint.ParameterValue {
+	var result map[string]*blueprint.ParameterValue
+	// the string has been validated by the schema, therefore the error is ignored here, since it will never happen.
+	_ = json.Unmarshal([]byte(input), &result)
+	return result
+}
+
+func expandArmBlueprintAssignmentResourceGroups(input string) map[string]*blueprint.ResourceGroupValue {
+	var result map[string]*blueprint.ResourceGroupValue
+	// the string has been validated by the schema, therefore the error is ignored here, since it will never happen.
+	_ = json.Unmarshal([]byte(input), &result)
+	return result
+}
+
+func expandArmBlueprintAssignmentIdentity(input []interface{}) (*blueprint.ManagedServiceIdentity, error) {
+	if len(input) == 0 {
```

Comment:
We'll want to confirm that input[0] isn't nil here
```suggestion
	if len(input) == 0  || input[0] == nil {
```